File: android/core/src/main/kotlin/com/slipstream/core/files/FileBrowser.kt
package com.slipstream.core.files

import java.io.File

/** One entry in a [DirectoryListing]. */
data class FileEntry(
    val name: String,
    val size: Long,
    val mtimeMs: Long,
    val isDirectory: Boolean,
    val mime: String?,
)

/** A capped directory listing. [truncated] is true when the directory held more than the cap. */
data class DirectoryListing(
    val entries: List<FileEntry>,
    val truncated: Boolean,
)

/** Lists a single directory's immediate children for the remote file browser. */
object FileBrowser {

    const val MAX_ENTRIES = 5000

    private val mimeByExtension = mapOf(
        "jpg" to "image/jpeg",
        "jpeg" to "image/jpeg",
        "png" to "image/png",
        "gif" to "image/gif",
        "webp" to "image/webp",
        "bmp" to "image/bmp",
        "mp4" to "video/mp4",
        "mov" to "video/quicktime",
        "mkv" to "video/x-matroska",
        "avi" to "video/x-msvideo",
        "mp3" to "audio/mpeg",
        "wav" to "audio/wav",
        "flac" to "audio/flac",
        "ogg" to "audio/ogg",
        "pdf" to "application/pdf",
        "txt" to "text/plain",
        "md" to "text/markdown",
        "json" to "application/json",
        "xml" to "application/xml",
        "html" to "text/html",
        "htm" to "text/html",
        "csv" to "text/csv",
        "zip" to "application/zip",
        "7z" to "application/x-7z-compressed",
        "rar" to "application/vnd.rar",
        "doc" to "application/msword",
        "docx" to "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xls" to "application/vnd.ms-excel",
        "xlsx" to "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "ppt" to "application/vnd.ms-powerpoint",
        "pptx" to "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "apk" to "application/vnd.android.package-archive",
    )

    /**
     * Lists the immediate children of [directory]: directories first, then files, each group
     * sorted case-insensitively by name. Caps the result at [MAX_ENTRIES]; when the directory
     * has more than that, [DirectoryListing.truncated] is true and only the first [MAX_ENTRIES]
     * (by the same sort order) are returned.
     */
    fun list(directory: File): DirectoryListing {
        val children = directory.listFiles()?.toList() ?: emptyList()
        val sorted = children.sortedWith(
            compareByDescending<File> { it.isDirectory }.thenBy { it.name.lowercase() },
        )
        val truncated = sorted.size > MAX_ENTRIES
        val capped = if (truncated) sorted.subList(0, MAX_ENTRIES) else sorted

        val entries = capped.map { f ->
            val isDir = f.isDirectory
            FileEntry(
                name = f.name,
                size = if (isDir) 0L else f.length(),
                mtimeMs = f.lastModified(),
                isDirectory = isDir,
                mime = if (isDir) null else mimeFor(f.name),
            )
        }
        return DirectoryListing(entries, truncated)
    }

    private fun mimeFor(name: String): String {
        val ext = name.substringAfterLast('.', missingDelimiterValue = "").lowercase()
        return mimeByExtension[ext] ?: "application/octet-stream"
    }
}

File: windows/src/Slipstream.Core/Files/FileBrowser.cs
namespace Slipstream.Core.Files;

public sealed record FileEntry(
    string Name,
    string Path,
    long Size,
    DateTimeOffset Modified,
    bool IsDirectory,
    string? Mime,
    string? ThumbnailToken);

public sealed record ListResult(string Path, IReadOnlyList<FileEntry> Entries, bool Truncated);

/// <summary>
/// Spec §6. Listings cap at 5000 entries with an honest flag rather than silently
/// truncating — no real directory on either device approaches the cap.
/// </summary>
public sealed class FileBrowser
{
    public const int MaxEntries = 5000;

    private static readonly Dictionary<string, string> MimeByExtension = new(StringComparer.OrdinalIgnoreCase)
    {
        [".mp4"] = "video/mp4", [".mkv"] = "video/x-matroska", [".avi"] = "video/x-msvideo",
        [".mov"] = "video/quicktime", [".webm"] = "video/webm", [".m4v"] = "video/x-m4v",
        [".mp3"] = "audio/mpeg", [".flac"] = "audio/flac", [".wav"] = "audio/wav",
        [".m4a"] = "audio/mp4", [".ogg"] = "audio/ogg", [".opus"] = "audio/opus",
        [".jpg"] = "image/jpeg", [".jpeg"] = "image/jpeg", [".png"] = "image/png",
        [".gif"] = "image/gif", [".webp"] = "image/webp", [".heic"] = "image/heic",
        [".pdf"] = "application/pdf", [".txt"] = "text/plain", [".zip"] = "application/zip",
    };

    public ListResult List(string path, string sort = "name")
    {
        if (!Directory.Exists(path))
            throw new DirectoryNotFoundException($"No folder at {path}.");

        var entries = new List<FileEntry>();
        var truncated = false;

        foreach (var entryPath in Directory.EnumerateFileSystemEntries(path))
        {
            if (entries.Count >= MaxEntries) { truncated = true; break; }

            var entry = Describe(entryPath);
            if (entry is not null) entries.Add(entry);
        }

        return new ListResult(path, Sort(entries, sort), truncated);
    }

    public FileEntry? Stat(string path) =>
        File.Exists(path) || Directory.Exists(path) ? Describe(path) : null;

    public IReadOnlyList<FileEntry> Roots() =>
        DriveInfo.GetDrives()
            .Where(d => d.IsReady)
            .Select(d => new FileEntry(
                string.IsNullOrWhiteSpace(d.VolumeLabel) ? d.Name : $"{d.VolumeLabel} ({d.Name})",
                d.RootDirectory.FullName, 0, DateTimeOffset.MinValue, true, null, null))
            .ToList();

    private static FileEntry? Describe(string path)
    {
        try
        {
            if (Directory.Exists(path))
            {
                var info = new DirectoryInfo(path);
                return new FileEntry(info.Name, info.FullName, 0,
                    new DateTimeOffset(info.LastWriteTimeUtc, TimeSpan.Zero), true, null, null);
            }

            var file = new FileInfo(path);
            return new FileEntry(file.Name, file.FullName, file.Length,
                new DateTimeOffset(file.LastWriteTimeUtc, TimeSpan.Zero), false,
                MimeFor(file.Extension), null);
        }
        catch (UnauthorizedAccessException)
        {
            return null; // Skip what we cannot read rather than failing the listing.
        }
        catch (IOException)
        {
            return null;
        }
    }

    private static string MimeFor(string extension) =>
        MimeByExtension.GetValueOrDefault(extension, "application/octet-stream");

    private static List<FileEntry> Sort(List<FileEntry> entries, string sort)
    {
        // Directories always lead, whatever the sort — that is how a file browser reads.
        var comparer = sort switch
        {
            "size" => (Comparison<FileEntry>)((a, b) => b.Size.CompareTo(a.Size)),
            "modified" => (a, b) => b.Modified.CompareTo(a.Modified),
            _ => (a, b) => string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase),
        };

        entries.Sort((a, b) =>
            a.IsDirectory != b.IsDirectory ? (a.IsDirectory ? -1 : 1) : comparer(a, b));

        return entries;
    }
}
